src/Build/BackEnd/Components/Communications/NodeProviderOutOfProc.cs

```diff
@@ -132,12 +144,7 @@ public void SendData(int nodeId, INodePacket packet)
         public void ShutdownConnectedNodes(bool enableReuse)
         {
             // Send the build completion message to the nodes, causing them to shutdown or reset.
-            List<NodeContext> contextsToShutDown;
-
-            lock (_nodeContexts)
-            {
-                contextsToShutDown = new List<NodeContext>(_nodeContexts.Values);
-            }
+            var contextsToShutDown = new List<NodeContext>(_nodeContexts.Values);
```

Comment:
Does this introduce a possible race between a call to ShutdownConnectedNodes and a call to create new nodes? Though it doesn't look like adding to `_nodeContexts` was locked before when adding so I guess that's not new.

Reply:
It is hard to find all possible race conditions here along with their probablility and consequences. I have found at least one race conditions: `ShutdownConnectedNodes` is in call stack of CTRL+C console cancellation so it can indeed happen anytime.

However, these changes did not it any worse than it was before. It is slightly better as  `_nodeContexts` is `ConcurentDictionary` now so it is guaranteed to be in consistent state when race condition happen, as oppose to former code when we could have read Values while it was in middle of adding.

I recommend to NOT invest more into this race condition.